### ISSUE TITLE
The output label is not match to the character in picture

### ISSUE THREAD

**Issue by myyang19770915** [keras-team/keras-io#514]
https://github.com/keras-team/keras-io/blob/0da4523f083ca68d4d1970efb523b0c83fac29ab/examples/vision/captcha_ocr.py#L176

When I run the code, the output plt.show() does not correct . Label is not match to the img.
I try to verify the code :  
          label = tf.strings.reduce_join(num_to_char(labels[i]+1)).numpy().decode("utf-8")
then it runs ok!

**Comment by BadMachine** [comment]
@myyang19770915 did u solve this?

**Comment by myyang19770915** [comment]
use this instead~
label = tf.strings.reduce_join(num_to_char(labels[i]+1)).numpy().decode("utf-8")

**Comment by BadMachine** [comment]
@myyang19770915 yeah, but we also have to replace this:
```
res = tf.strings.reduce_join(num_to_char(res)).numpy().decode("utf-8")

```
to this:
```
res = tf.strings.reduce_join(num_to_char(res+1)).numpy().decode("utf-8")
```

**Comment by NickCis** [comment]
I've also came across the same issue, it seems that there was a change in behaviour on how `layers.experimental.preprocessing.StringLookup` works. As `char_to_num` is using `num_oov_indices=0` while `num_to_char` is using the default one (which happens to be `num_oov_indices=1`) all mappings seem to be drifted by 1.

The easiest way to apply a quick fix is to wrap the `num_to_char` function in order to perform that drift:
```python3
# Mapping integers back to original characters
_num_to_char = layers.experimental.preprocessing.StringLookup(
    vocabulary=char_to_num.get_vocabulary(), mask_token=None, invert=True
)

num_to_char = lambda x: _num_to_char(x + tf.ones(tf.shape(x), dtype='int64')
```

---

I've tried changing the `num_oov_indices=0` of `char_to_num` (erasing this parameter, so it takes the default argument which will be the same as `num_to_char`):

```python3
# Mapping characters to integers
char_to_num = layers.experimental.preprocessing.StringLookup(
    vocabulary=list(characters), mask_token=None
)

# Mapping integers back to original characters
num_to_char = layers.experimental.preprocessing.StringLookup(
    vocabulary=char_to_num.get_vocabulary(), mask_token=None, invert=True
)
```

This fixes the mapping but training will fail with the following error:
```
Epoch 1/100
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-31-2c137061a23a> in <module>()
     11     validation_data=validation_dataset,
     12     epochs=epochs,
---> 13     callbacks=[early_stopping],
     14 )

6 frames
/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/execute.py in quick_execute(op_name, num_outputs, inputs, attrs, ctx, name)
     58     ctx.ensure_initialized()
     59     tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,
---> 60                                         inputs, attrs, num_outputs)
     61   except core._NotOkStatusException as e:
     62     if name is not None:

InvalidArgumentError: 2 root error(s) found.
  (0) Invalid argument:  Saw a non-null label (index >= num_classes - 1) following a null label, batch: 10 num_classes: 54 labels: 30,9,30,53,5 labels seen so far: 30,9,30
	 [[node ocr_model_v1/ctc_loss/CTCLoss (defined at <ipython-input-30-de6e4835d664>:17) ]]
  (1) Invalid argument:  Saw a non-null label (index >= num_classes - 1) following a null label, batch: 10 num_classes: 54 labels: 30,9,30,53,5 labels seen so far: 30,9,30
	 [[node ocr_model_v1/ctc_loss/CTCLoss (defined at <ipython-input-30-de6e4835d664>:17) ]]
	 [[gradient_tape/ocr_model_v1/ctc_loss/Shape/_88]]
0 successful operations.
0 derived errors ignored. [Op:__inference_train_function_71434]

Function call stack:
train_function -> train_function
```

Any idea about this?.

**Comment by NickCis** [comment]
Well, I managed to finally understand what was failing.

As I've said in the previous message the wrong translation was due to the difference between `char_to_num` and `num_to_char`. The first one was created without an out of vocabulary character (`num_oov_indices=0`) while the latter with one. Apparently, this didn't create a problem in previous tensorflow versions (i.e.: `2.4.x`, I've been manually installing `2.4.3` in colab in order to get the previous behavior), but, with newer ones (`2.5.x`, at the time of writing colab uses `2.5.0`) it drifts numbers by one.

Following the reasoning of the previous message, if both mappings are created with the default `num_oov_indices` (which is having one out of vocabulary), translation works fine (ie.: `char_to_num(num_to_char(tf.strings.unicode_split('test', input_encoding="UTF-8")))` will yield `test`):

```python3
# Mapping characters to integers
char_to_num = layers.experimental.preprocessing.StringLookup(
    vocabulary=list(characters), mask_token=None
)

# Mapping integers back to original characters
num_to_char = layers.experimental.preprocessing.StringLookup(
    vocabulary=char_to_num.get_vocabulary(), mask_token=None, invert=True
)
```

But in order the CTC layer to work correctly, the last dense layer has to have `len(char_to_num.vocabulary()) + 1` units. The original code was using `len(character)+1` units for this layer, as now we have added _oov_,  `len(characters) != len(char_to_num.vocabulay())`. If I'm not misunderstanding something this is due to the _blank pseudo character_ introduced in order to perform the ctc loss (the problem of encoding a repeated character).

```python3
    # Output layer
    x = layers.Dense(len(char_to_num.get_vocabulary()) + 1, activation="softmax", name="dense2")(x)
```

I've created the pr in order to fix the example ( #533 ) and also you can checked it on [this colab](https://colab.research.google.com/drive/1kSSDhSofz1jNaxQr8f3Ex7v9JiF2GVA9?usp=sharing)

**Comment by NickCis** [comment]
This should be fixed by #533 . The issue can be closed.

**Comment by myyang19770915** [comment]
It is ok, thanks

取得 iOS 版 Outlook<https://aka.ms/o0ukef>
________________________________
寄件者: SuryanarayanaY ***@***.***>
寄件日期: Thursday, August 3, 2023 12:45:57 AM
收件者: keras-team/keras-io ***@***.***>
副本: myyang19770915 ***@***.***>; Mention ***@***.***>
主旨: Re: [keras-team/keras-io] The output label is not match to the character in picture (#514)


Hi @myyang19770915<https://github.com/myyang19770915> ,

The Proposed PR merged and I have tested the tutorial with Tf2.12v and its working fine. Please refer the attached gist<https://colab.sandbox.google.com/gist/SuryanarayanaY/b4d5420c9994d3fe23f3dfda10a9b48c/kerasio_514-kerasio-captcha_ocr-r2.ipynb>.

Shall we mark it as resolved. Please check and confirm. Thanks!
